### PR TITLE
Story/vspc 172

### PR DESCRIPTION
**Pull request by bnayak4**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

This is pradeeps' story and it consists of changes of VSPC-172,VSPC-196 and VSPC-197

### CODE REVIEW

**Comment by jdamerow** [comment]
Codefactor failed.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java [hunk 1]
```diff
@@ -36,37 +37,55 @@ public abstract class LinkManager<L extends ILink<T>, T extends IVSpaceElement,
...
         T target = getTarget(linkedId);
         link.setName(linkLabel);
+        link.setDescription(linkDesc);
         link.setTarget(target);
         U displayLink = createDisplayLink(link);
-        setDisplayProperties(displayLink, id, positionX, positionY, rotation, displayType, linkImage, imageFilename);
+        if(existingImageId!=null && !existingImageId.trim().isEmpty()) {
+            setDisplayProperties(displayLink, positionX, positionY, rotation, displayType, existingImageId);
+        }
+        else {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java:59]: please follow our style guide

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java [hunk 2]
```diff
@@ -36,37 +37,55 @@ public abstract class LinkManager<L extends ILink<T>, T extends IVSpaceElement,
...
         link.setName(title);
+        link.setDescription(linkDesc);
         link.setTarget(target);
         U displayLink = getDisplayLink(linkDisplayId);
-        setDisplayProperties(displayLink, id, positionX, positionY, rotation, displayType, linkImage, imageFilename);
-        return updateLinkAndDisplay(link, displayLink);
+        if(existingImageId!=null && !existingImageId.trim().isEmpty()) {
+            setDisplayProperties(displayLink, positionX, positionY, rotation, displayType, existingImageId);
+        }
+        else {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java:83]: same here, code style

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java [hunk 3]
```diff
@@ -100,15 +119,24 @@ protected void validateSpace(String id) throws SpaceDoesNotExistException {
...
+    }
+    
+    protected void setLinkDisplay(ILinkDisplay linkDisplay, float positionX, float positionY, int rotation, DisplayType displayType) {
+    	linkDisplay.setPositionX(positionX);
+        linkDisplay.setPositionY(positionY);
+        linkDisplay.setRotation(rotation);
+        linkDisplay.setType(displayType != null ? displayType : DisplayType.ARROW);
+    	
+    	
+
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java:131]: too many empty lines

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java [hunk 4]
```diff
@@ -36,37 +37,55 @@ public abstract class LinkManager<L extends ILink<T>, T extends IVSpaceElement,
...
         return updateLinkAndDisplay(link, displayLink);
 
     }
 
     @Override
     public U updateLink(String title, String id, float positionX, float positionY, int rotation, String linkedId,
-            String linkLabel, String linkId, String linkDisplayId, DisplayType displayType, byte[] linkImage,
-            String imageFilename)
-            throws SpaceDoesNotExistException, LinkDoesNotExistsException, ImageCouldNotBeStoredException {
+            String linkLabel, String linkDesc,String linkId,  String linkDisplayId, DisplayType displayType, byte[] linkImage,
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java:68]: there should be one space between `,String` and only one space after the comma (`,  String`)

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java [hunk 5]
```diff
@@ -56,7 +59,8 @@ public ResponseEntity<String> createExternalLink(@PathVariable("id") String id,
         ExternalLinkDisplayMode externalLinkOpenMode = howToOpen.isEmpty() ? null
                 : ExternalLinkDisplayMode.valueOf(howToOpen);
         IExternalLinkDisplay display = externalLinkManager.createLink(title, id, new Float(x), new Float(y), 0,
-                externalLink, title, type, linkImage, filename, externalLinkOpenMode);
+                externalLink, title,desc, type, linkImage, filename, externalLinkOpenMode, imageId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java:62]: blank missing: `title,desc`

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddModuleLinkController.java [hunk 6]
```diff
@@ -33,12 +33,14 @@ public class AddModuleLinkController {
...
-            @RequestParam("moduleLinkImage") MultipartFile file)
-            throws NumberFormatException, SpaceDoesNotExistException, IOException, ImageCouldNotBeStoredException {
+	public ResponseEntity<String> createModuleLink(@PathVariable("id") String id, @RequestParam("x") String x,
+			@RequestParam("y") String y, @RequestParam("rotation") String rotation,
+			@RequestParam("moduleLinkLabel") String title, @RequestParam("linkedModule") String linkedModuleId,
+			@RequestParam("moduleLinkLabel") String moduleLinkLabel,
+			@RequestParam("moduleLinkDesc") String moduleLinkDesc, @RequestParam("moduleType") String displayType,
+			@RequestParam("moduleLinkImage") MultipartFile file,
+			@RequestParam(value = "imageId", required = false) String imageId)
+			throws NumberFormatException, SpaceDoesNotExistException, IOException, ImageCouldNotBeStoredException, ImageDoesNotExistException {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddModuleLinkController.java:43]: Indentation

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddModuleLinkController.java [hunk 7]
```diff
@@ -63,7 +65,7 @@ public ResponseEntity<String> createModuleLink(@PathVariable("id") String id, @R
         IModuleLinkDisplay display;
         try {
             display = moduleLinkManager.createLink(title, id, new Float(x), new Float(y), new Integer(rotation),
-                    linkedModuleId, moduleLinkLabel, type, linkImage, filename);
+					linkedModuleId, moduleLinkLabel, moduleLinkDesc, type, linkImage, filename, imageId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddModuleLinkController.java:68]: why is the indentation changed here?

**Review by jdamerow — CHANGES_REQUESTED** [review]
there are a bunch of wrong indented lines introduced here, please fix

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Tests failed:
```
test_updateLink_success(edu.asu.diging.vspace.core.services.impl.ExternalLinkManagerTest)  Time elapsed: 0.038 sec  <<< ERROR!
java.lang.StackOverflowError
	at edu.asu.diging.vspace.core.services.impl.ExternalLinkManager.updateLink(ExternalLinkManager.java:156)
```

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 8]
```diff
@@ -2350,6 +3021,7 @@ <h6 class="alert-heading">
                                 class="form-control-sm modulelink-targetEdit">
                                 <option th:value="IMAGE">Image</option>
                                 <option th:value="ARROW" th:selected>Link</option>
+                                <option th:value="ARROW">Link</option>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:3024]: why is this here twice?

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 9]
```diff
@@ -2313,11 +2974,13 @@ <h6 class="alert-heading">
                         </div>
                     </div>
                     <input type="hidden" name="moduleLinkIdValueEdit" id="moduleLinkIdValueEdit" /> 
+                    <input type="hidden" name="moduleLinkImageIdEdit" id="moduleLinkImageIdEdit" />
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:2977]: what is this for?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditModuleLinkController.java [hunk 10]
```diff
@@ -33,14 +33,19 @@ public ResponseEntity<String> editModuleLink(@PathVariable("id") String id, @Req
...
-            @RequestParam("moduleLinkImage") MultipartFile file) throws NumberFormatException,
-            SpaceDoesNotExistException, LinkDoesNotExistsException, IOException, ImageCouldNotBeStoredException {
+            @RequestParam("moduleLinkImage") MultipartFile file, @RequestParam(value="imageId", required=false) String imageId) throws NumberFormatException,
+            SpaceDoesNotExistException, LinkDoesNotExistsException, IOException, ImageCouldNotBeStoredException, ImageDoesNotExistException {
 
         ResponseEntity<String> validation = checkIfSpaceExists(spaceManager, id, x, y);
         if (validation != null) {
             return validation;
+        }else if(imageId==null || imageId.equals("")){
+            String[] token = moduleLinkImageIdEdit.split("/");
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditModuleLinkController.java:47]: what is `moduleLinkImageIdEdit` and why does it need to be split?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java [hunk 11]
```diff
@@ -30,33 +31,35 @@ public class EditSpaceLinkController extends EditSpaceLinksController {
...
+	public ResponseEntity<String> editSpaceLink(@PathVariable("id") String id, @RequestParam("x") String x,
+			@RequestParam("y") String y, @RequestParam("rotation") String rotation,
+			@RequestParam("spaceLinkLabel") String title, @RequestParam("linkedSpace") String linkedSpaceId,
+			@RequestParam("spaceLinkLabel") String spaceLinkLabel,@RequestParam("spaceLinkDesc") String spaceLinkDesc,
+			@RequestParam("spaceLinkIdValueEdit") String spaceLinkIdValueEdit,
+			@RequestParam("spaceLinkDisplayId") String spaceLinkDisplayId,
+			@RequestParam(value = "spaceLinkImageIdEdit", required = false) String spaceLinkImageIdEdit,
+			@RequestParam("type") String displayType, @RequestParam("spaceLinkImage") MultipartFile file,
+			@RequestParam(value = "imageId", required = false) String imageId) throws NumberFormatException,
+			SpaceDoesNotExistException, LinkDoesNotExistsException, IOException, ImageCouldNotBeStoredException, ImageDoesNotExistException {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java:43]: indentation of this whole block is too much

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 12]
```diff
@@ -909,6 +980,321 @@
...
+        		templateSelection: formatSelectionEdit
+            });
+        	
+        	function formatImage (image) {
+            	if (!image.id) {
+            		return image.text;
+            	}
+            	var tbaseUrl = [[@{|/api/image/|}]];
+            	var $image = $(
+            			'<span><img src="' + tbaseUrl + image.id + '" class="img-thumbnail" /> ' + image.text + '</span>'
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1061]: indentation

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java [hunk 13]
```diff
@@ -609,6 +614,24 @@ public void updateContentOrder(List<ContentBlock> contentBlockList) throws Block
...
+    @Override
+    public void updateContentOrder(List<ContentBlock> contentBlockList) throws BlockDoesNotExistException {
+        if (contentBlockList == null || contentBlockList.isEmpty()) {
+            return;
+        }
+        
+        // Get the slide from the first content block
+        ISlide slide = contentBlockList.get(0).getSlide();
+        updateContentOrder(contentBlockList, slide);
+    }
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java]: why is this method needed? why not just modifying the existing one? It's only called once, right?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java [hunk 14]
```diff
@@ -38,16 +39,26 @@ public class AddExternalLinkController {
...
         ISpace space = spaceManager.getSpace(id);
+        String desc = "";
         if (space == null) {
             return new ResponseEntity<>("{'error': 'Space could not be found.'}", HttpStatus.NOT_FOUND);
         }
 
         byte[] linkImage = null;
         String filename = null;
+        
+        if (file == null && imageId == null) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java]: not every link needs to have an image, e.g. if they use the default icon. Won't this break any other case?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/StaffSearchSpaceController.java [hunk 15]
```diff
@@ -22,6 +26,9 @@ public class StaffSearchSpaceController {
 
     @Autowired
     private IStaffSearchManager staffSearchManager;
+    
+    @Autowired
+    private ISpaceManager spaceManager;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/StaffSearchSpaceController.java]: is this used anywhere? are any of the changes in this file needed?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateContentOrderController.java [hunk 16]
```diff
@@ -23,17 +30,25 @@ public class UpdateContentOrderController {
...
 
+        ObjectMapper objectMapper = new ObjectMapper();
+        ISlide slide = slideManager.getSlide(slideId);
+        List<ContentBlock> contentBlockList;
         try {
-            contentBlockManager.updateContentOrder(contentBlockList);
-        } catch (BlockDoesNotExistException e) {
-            logger.warn("Block Id does not exist, bad request.", e);
-            return new ResponseEntity<List<ContentBlock>>(HttpStatus.BAD_REQUEST);
+            System.out.println(contentBlockString);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateContentOrderController.java]: remove

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateContentOrderController.java [hunk 17]
```diff
@@ -23,17 +30,25 @@ public class UpdateContentOrderController {
...
+            contentBlockList = objectMapper.readValue(contentBlockString, new TypeReference<List<ContentBlock>>(){});
+            contentBlockManager.updateContentOrder(contentBlockList,slide);
+        } catch (Exception e) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateContentOrderController.java:49]: exceptions should be specific or is `Exception` really thrown here (in this case, there should be a comment here).

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateContentOrderController.java [hunk 18]
```diff
@@ -23,17 +30,25 @@ public class UpdateContentOrderController {
...
+            logger.warn("Error while parsing content blocks.", e);
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/UpdateContentOrderController.java:53]: when is this thrown? does it always indicate a bad request?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/StaffSearchSpaceController.java (hunk 15 above)
Comment (Girik1105) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/StaffSearchSpaceController.java]: Removed the unused ISpaceManager dependency and unused imports.

**Pull request closed without merging**